### ISSUE TITLE
Assertion triggered when a serialized parallel region is after a detached task

### ISSUE THREAD

**Issue by shiltian** [llvm/llvm-project#50602]
|  |  |
| --- | --- |
| Bugzilla Link | [51258](https://llvm.org/bz51258) |
| Version | unspecified |
| OS | All |
| Blocks | llvm/llvm-project#49160, llvm/llvm-project#53081  |

## Extended Description 
The reproducer is every simple:
```
#include <omp.h>

int main(int argc, char *argv[]) {
#pragma omp parallel
  {
    omp_event_handle_t event;
    int a = 0;

#pragma omp task shared(a) detach(event)
    { a = 1; }

#pragma omp parallel
    { a = 2; }

#pragma omp taskwait
  }
  return 0;
}
```
Please ignore that the detached task is never fulfilled, because it is not the point. If it is executed with env `OMP_NUM_THREADS=1`, then an assertion is triggered.
```
Assertion failure at kmp_runtime.cpp(1114): serial_team->t.t_task_team[this_thr->th.th_task_state] == __null.
OMP: Error #&#8203;13: Assertion failure at kmp_runtime.cpp(1114).
OMP: Hint Please submit a bug report with this message, compile and run commands used, and machine configuration info including native compiler and operating system versions. Faster response will be obtained by including all program sources. For information on submitting this issue, please see https://bugs.llvm.org/.
```
Basically, `__kmp_serialized_parallel` assumes that the corresponding task team should be nullptr. However, if a detached task is encountered before, the task team will be setup in `__kmp_task_alloc`.

**Comment by shiltian** [comment]
@AndreyChurbanov Currently it hits the assertion in `__kmp_serialized_parallel`:
```
  if (__kmp_tasking_mode != tskm_immediate_exec) {
    KMP_DEBUG_ASSERT(
        this_thr->th.th_task_team ==
        this_thr->th.th_team->t.t_task_team[this_thr->th.th_task_state]);
    KMP_DEBUG_ASSERT(serial_team->t.t_task_team[this_thr->th.th_task_state] ==
                     NULL);
    KA_TRACE(20, ("__kmpc_serialized_parallel: T#%d pushing task_team %p / "
                  "team %p, new task_team = NULL\n",
                  global_tid, this_thr->th.th_task_team, this_thr->th.th_team));
    this_thr->th.th_task_team = NULL;
  }
```
If we don't set `this_thr->th.th_task_team = NULL;` here, later when the nested parallel region exits, in `__kmp_task_team_wait`, the task team will still be reset, leading to another issue that when the outer parallel region exits, another assertion `master_th->th.th_task_team == team->t.t_task_team[master_th->th.th_task_state]` will be hit.

Can we switch the `th_task_state` to another state in `__kmp_serialized_parallel` if the task team has detached task encountered? Or can we use the state in this case?

**Comment by shiltian** [comment]
Add @jprotze as well.

**Comment by AndreyChurbanov** [comment]
Looks like we need to get rid of "optimization" of re-usage of serial team in nested serial parallel regions.  Switching task teams may not help if we would have 3 or 4 nested parallel regions, as we only have two task teams those supposed to be used sequentially one after the other.  My view: the proper solution should be to "stack" serial teams, like we do in case of e.g. nesting of serial - active - serial regions.  We need to do the same for nesting of serial - serial regions. Then the problem with task teams should be solved automatically.
BTW, your test reveals some other problem as well, as it crashes sometimes in case of active outer parallel region.
Will try to work on this soon.

**Comment by shiltian** [comment]
@AndreyChurbanov Thanks for the info. I'll fix it according to your opinion.

**Comment by AndreyChurbanov** [comment]
Another problem appeared to be my fault, - I've broken the test replacing taskwait with omp_fulfill_event call, and got stack corruption because of local variable access out of scope.  So the only problem is serial team re-use, that causes problems with task team (besides other complications, like dispatch buffers needed to be stacked, etc.). Initial design didn't suppose to use task team in serialized parallel.

Thanks for taking care of this.

**Comment by shiltian** [comment]
I'm not sure I still have the capacity to work on this, but it's on my list. Remove myself for now in case any one else is interested of carrying it on.
